tests/framework/CodeInterfaceTests/PARCS/constraints.py

```diff
@@ -0,0 +1,100 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# @ author: Mohammad Abdo (@Jimmy-INL)
+
+import numpy as np
+
+def constrain(Input):#Complete this: give the function the correct name#
+  """
+    This fuction calls the explicit constraint whose name is passed through Input.name
+    the evaluation function g is negative if the explicit constraint is violated and positive otherwise.
+    This suits the constraint handling in the Genetic Algorithms,
+    but not the Gradient Descent as the latter expects True if the soltution passes the constraint and False if it violates it.
+    @ In, Input, object, RAVEN container
+    @ Out, g, float, expilicit constraint evaluation (negative if violated and positive otherwise)
+  """
+  g = eval(Input.name)(Input)
+  return g
+
+def implicitConstraint(Input):
+  """
+    Evaluates the implicit constraint function at a given point/solution ($\vec(x)$)
+    @ In, Input, object, RAVEN container
+    @ Out, g(inputs x1,x2,..,output or dependent variable), float, implicit constraint evaluation function
+            the way the constraint is designed is that
+            the constraint function has to be >= 0,
+            so if:
+            1) f(x,y) >= 0 then g = f
+            2) f(x,y) >= a then g = f - a
+            3) f(x,y) <= b then g = b - f
+            4) f(x,y)  = c then g = 1e-6 - abs((f(x,y) - c)) (equality constraint)
+  """
+  g = eval(Input.name)(Input)
+  return g
+
+
+def expConstr1(Input):#You are free to pick this name but it has to be similar to the one in the xml#
+  """
+    Let's assume that the constraint is:
+    $ x3+x4 < 8 $
+    then g the constraint evaluation function (which has to be > 0) is taken to be:
+    g = 8 - (x3+x4)
+    in this case if g(\vec(x)) < 0 then this x violates the constraint and vise versa
+    @ In, Input, object, RAVEN container
+    @ out, g, float, explicit constraint 1 evaluation function
+  """
+  # add the first explicit constraint here in case you have any
+  pass
+
+def expConstr2(Input):
+  """
+    Explicit Equality Constraint:
+    let's consider the constraint x1**2 + x2**2 = 25
+    The way to write g is to use a very small number for imnstance, epsilon = 1e-12
+    and then g = epsilon - abs(constraint)
+    @ In, Input, object, RAVEN container
+    @ out, g, float, explicit constraint 2 evaluation function
+  """
+  # add the second explicit constraint here in case you have any
+  pass
+
+def impConstr1(Input):
+  """
+    The implicit constraint involves variables from the output space, for example the objective variable or
+    a dependent variable that is not in the optimization search space
+    @ In, Input, object, RAVEN container
+    @ out, g, float, implicit constraint 1 evaluation function
+  """
+  g = max(0,(1300 - Input.max_boron))
```

Comment:
@Nhatkhang, this is wrong, the fitness itself should take care of the max thing. Please revert it back to what I have sent you. g should be positive if it satisfies the constraint and negative if it violates, then the fitness will add (max(0,-g)) to penalize the violation so if g is positive it will add zero if g is negative it will be penalized. Please take a look at the fitness definition in the manual for more details.

Reply:
I see, thank you. Then the fitness function should be the same as what we did in MOF at NCSU

Reply:
In GA, we have multiple versions of fitness: InvLin and feasibleFirst. the way I have sent you the constraints will render the fitness exactly similar to what you had in the slides.

Reply:
only modify the names of the variables and it should be ready to go.